homeassistant/components/nextcloud/sensor.py

```diff
@@ -0,0 +1,170 @@
+    "storage_num_fed_shares_received": [
+        "nextcloud",
+        "shares",
+        "num_fed_shares_received",
+    ],
+    "server_webserver": ["server", "webserver"],
+    "server_php_version": ["server", "php", "version"],
+    "server_php_memory_limit": ["server", "php", "memory_limit"],
+    "server_php_max_execution_time": ["server", "php", "max_execution_time"],
+    "server_php_max_file_size": ["server", "php", "upload_max_filesize"],
+    "server_database_type": ["server", "database", "type"],
+    "server_database_version": ["server", "database", "version"],
+    "server_database_size": ["server", "database", "size"],
+    "active_users_last_5_minutes": ["activeUsers", "last5minutes"],
+    "active_users_last_1_hour": ["activeUsers", "last1hour"],
+    "active_users_last_24_hours": ["activeUsers", "last24hours"],
+}
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None) -> None:
+    """Set up the Nextcloud sensor."""
+    # Fetch Nextcloud Monitor api data
+    ncm = NextcloudMonitor(
+        config[CONF_URL], config[CONF_USERNAME], config[CONF_PASSWORD]
+    )
+
+    hass.data[DOMAIN] = ncm.data
+
+    def nextcloud_update(event_time):
+        """Update data from nextcloud api."""
+        ncm.update()
+        hass.data[DOMAIN] = ncm.data
+
+    # Update sensors on time interval
+    track_time_interval(hass, nextcloud_update, config[CONF_SCAN_INTERVAL])
+
+    # Create list of sensors based on available nextcloud api data
+    sensors = []
+    for name, dict_path in NEXTCLOUD_ITEMS_EXPECTED.items():
+        sensors.append(NextcloudSensor(name, dict_path))
+
+    # Setup sensors
+    add_entities(sensors, True)
+
+
+class NextcloudSensor(Entity):
+    """Represents a Nextcloud sensor."""
+
+    def __init__(self, item, value_path):
+        """Initialize the Nextcloud sensor."""
+        self.item = item
+        self.value = None
+        self.value_path = value_path
+
+    @property
+    def icon(self):
+        """Return the icon for this sensor."""
+        if self.item[:6] == "system":
+            return "mdi:memory"
+        if self.item[:7] == "storage":
+            return "mdi:folder-account"
+        if self.item[:6] == "server":
+            return "mdi:server"
+        if self.item[:6] == "active":
+            return "mdi:account-multiple"
+        return "mdi:cloud"
+
+    @property
+    def name(self):
+        """Return the name for this sensor."""
+        return f"{DOMAIN}_{self.item}"
+
+    @property
+    def state(self):
+        """Return the state for this sensor."""
+        return self.value
+
+    @property
+    def unique_id(self):
+        """Return the unique ID for this sensor."""
+        return f"{DOMAIN}_{self.name}"
+
+    def update(self):
+        """Update the sensor."""
+        try:
+            data = self.hass.data[DOMAIN]
+            for path in self.value_path:
+                data = data[path]
```

Comment:
Aren't you overriding `data` with every run of the loop? Not quite sure what this loop does tbh, could you clarify my mistunderstanding? 😅

Reply:
There's probably a better way to do it but basically this loop is walking the dictionary to get to the relevant value.
`NEXTCLOUD_ITEMS_EXPECTED` contains a list of keys to follow in order to get to the relevant value for each 'sensor'. for example, the value for `system_version` exists at `hass.data[DOMAIN]['nextcloud']['system']['version']` so the logic is to look up `hass.data[DOMAIN]['nextcloud']` then `['system']` then `['version']`

Reply:
IMHO it would be better to outsource this into the python package and only call a function update_specific(x) which only retrieves the newest value for the specified `x` e.g. `system_version`.

Reply:
I originally had this but since there are so many sensors it was making so many calls to the API. Because the API returns all the data at once it seems most efficient to make a single call and then parse the data that the API returns.

Reply:
Yeah that is a good idea, however you could first call update_all_data() which returns nothing and then you could make a function get_data(x) that only retrieves the value for the specified x e.g. system_version. Therefore, outsourcing the filtering of the updated values to the python package.

Reply:
I guess in that case i would reduce `NEXTCLOUD_ITEMS_EXPECTED` to a list istead a dictionary of lists. Does that sound right to you?